src/usr/share/gpgmailer/gpgkeyring.py

```diff
@@ -0,0 +1,134 @@
+# These trust levels come from gnupg, u means ultimate, f means full, and m
+#   means marginal.
+valid_owner_trust_levels = ('u', 'f', 'm')
+
+# GpgKeyRing caches and checks validity, expiration, and trust for pgp keys.
+class GpgKeyRing:
+    # Loads the keyring and stores it in a dict.
+    #
+    # gnupg_home: the keyring directory for gnupg.
+    def __init__(self, gnupg_home):
+        self.logger = logging.getLogger('GpgKeyRing')
+        self.gnupg_home = gnupg_home
+        self.gpg = gnupg.GPG(gnupghome=self.gnupg_home)
+        self.fingerprint_to_key_dict = {}
+
+        for key in self.gpg.list_keys():
+            
+            # Key expiration dates are in Unix time. An expiration date of None
+            #   means that the key does not expire.
+            if key['expires'] == '':
+                key['expires'] = None
+            else:
+                key['expires'] = int(key['expires'])
+
+            self.fingerprint_to_key_dict[key['fingerprint']] = {
+                'expires': key['expires'],
+                'ownertrust': key['ownertrust'],
+                'fingerprint': key['fingerprint']
+            }
+
+
+    # Check if a key with the given fingerprint is still valid after the given date.
+    #
+    # fingerprint: the fingerprint of the key to check
+    # expiration_date: the date (in Unix time) to which to compare the key's
+    #   expiration date.
+    def is_current(self, fingerprint, expiration_date):
+        current = False
+        self._fingerprint_is_valid(fingerprint)
+
+        self.logger.trace('Checking expiration for key %s at date %s.' % (fingerprint,
+            expiration_date))
+
+        if ((self.fingerprint_to_key_dict[fingerprint]['expires'] == None) or 
+             (self.fingerprint_to_key_dict[fingerprint]['expires'] > expiration_date)):
+
+            current = True
+            self.logger.trace('Key %s is current.' % fingerprint)
+
+        else:
+            self.logger.trace('Key %s expires before date %s.' % (fingerprint, expiration_date))
+
+        return current
+
+
+    # Check if a key with the given fingerprint is trusted.
+    #
+    # fingerprint: the fingerprint of the key to check
+    def is_trusted(self, fingerprint):
+        trusted = False
+        self._fingerprint_is_valid(fingerprint)
+
+        if self.fingerprint_to_key_dict[fingerprint]['ownertrust'] in valid_owner_trust_levels:
+            trusted = True
+
+        else:
+            self.logger.warn('Key %s is not trusted' % fingerprint)
+
+        return trusted
+
+
+    # Looks up a key fingerprint and returns the expiration date if it exists,
+    #   throws an exception if it is invalid or not in the keyring. If a format
+    #   is specified, the date will be expressed in that format.
+    #
+    # fingerprint: A PGP key fingerprint
+    # date_format: A format string compatible with Python's strftime.
+    def get_key_expiration_date(self, fingerprint, date_format=None):
```

Comment:
I feel this method should return a date, not a formatted String. The formatting should be the responsibility of the consumer.